### PR TITLE
PWGHF: Adding first implementation of D0-D0bar correlation code (data/MC-reco/MC-kine)

### PR DESCRIPTION
**Pull request by fcolamar**
Implementation of analysis task for D0-D0bar correlations.
Allows running in data-like, MC-reco and MC-kine modes.
ULS and LS analysis enables.

### CODE REVIEW

**Comment by fcolamar** [comment]
Hi @vkucera, @ginnocen, I'm opening this draft pull request for the D0-D0bar tasks integration, let me know for any comment

**Review by vkucera — CHANGES_REQUESTED** [review]
Hi @fcolamar .
Please fix the naming so that it complies with the conventions.

**Comment by fcolamar** [comment]
Hi @vkucera, I modified the task names, which were inadequate, and the filename according to the coding conventions.
Let me know if there's some other renaming that I missed, thanks

**Comment by vkucera** [comment]
Thanks, please have a look at the [General Naming Rules](https://rawgit.com/AliceO2Group/CodingGuidelines/master/naming_formatting.html?showone=General_Naming_Rules#General_Naming_Rules). That might be clearer.

**Comment by ginnocen** [comment]
@vkucera but what is the specific problem in this case?

**Comment by vkucera** [comment]
> Thanks, please have a look at the [General Naming Rules](https://rawgit.com/AliceO2Group/CodingGuidelines/master/naming_formatting.html?showone=General_Naming_Rules#General_Naming_Rules). That might be clearer.

There are still some variables, histogram names and other pieces, like underscores, to fix. Also missing braces.
It might be better to fix those before starting with the code comments.

**Comment by vkucera** [comment]
Thanks for the update. It still does not comply with the conventions but I will comment together with other suggestions.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 1]
```diff
@@ -0,0 +1,823 @@
...
+double getDeltaPhi(double phiD, double phiDbar)
+{
+  double dPhi = phiDbar - phiD;
+
+  if (dPhi < -o2::constants::math::PI / 2.)
+    dPhi = dPhi + 2. * o2::constants::math::PI;
+  if (dPhi > 3. * o2::constants::math::PI / 2.)
+    dPhi = dPhi - 2. * o2::constants::math::PI;
+
+  return dPhi;
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: You can replace this with
```c++
return RecoDecay::constrainAngle(phiDbar - phiD, -o2::constants::math::PI / 2.);
```
See https://github.com/AliceO2Group/AliceO2/pull/5521

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 2]
```diff
@@ -0,0 +1,823 @@
...
+double evaluatePhiByVertex(double xVertex1, double xVertex2, double yVertex1, double yVertex2)
+{
+  double phi = std::atan2((yVertex2 - yVertex1), (xVertex2 - xVertex1));
+
+  if (phi < 0.)
+    phi = phi + 2. * o2::constants::math::PI;
+  if (phi > 2. * o2::constants::math::PI)
+    phi = phi - 2. * o2::constants::math::PI;
+
+  return phi;
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: You can replace this with
```c++
return RecoDecay::Phi(xVertex2 - xVertex1, yVertex2 - yVertex1);
```

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 3]
```diff
@@ -0,0 +1,823 @@
...
+}
+
+/// D0 analysis task - for real data and data-like analysis (i.e. reco-level w/o matching request via MC truth)
+struct TaskD0D0barCorrelation {
+
+  double maxEtaCut = 5., PtThresholdForMaxEtaCut = 10.; //for hDDbarVsEtaCut, gives eta increment of 0.1 and pt thr increments of 0.5
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: - Is `maxEtaCut` supposed to be different from `cutEtaCandMax`?
- Should `maxEtaCut`, `PtThresholdForMaxEtaCut` be configurable or constant?
- The increments are independent of these values. So the comment is incorrect.
- The actual increments of `ptCut` and `etaCut` are hard-coded, whereas the bin widths of `hDDbarVsEtaCut` are defined with separate binning factors. They should be linked via constants.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 4]
```diff
@@ -0,0 +1,823 @@
...
+  Filter filterSelectCandidates = (aod::hf_selcandidate_d0::isSelD0 >= d_selectionFlagD0 || aod::hf_selcandidate_d0::isSelD0bar >= d_selectionFlagD0bar);
+
+  void process(aod::Collision const& collision, soa::Filtered<aod::Big2Prong> const& candidates, aod::BigTracks const& tracks)
+  {
+    for (auto& candidate1 : candidates) {
+      if (cutEtaCandMax >= 0. && std::abs(candidate1.eta()) > cutEtaCandMax)
+        continue;
+      if (cutPtCandMin >= 0. && std::abs(candidate1.pt()) < cutPtCandMin)
+        continue;
+      //check decay channel flag for candidate1
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Add readability-braces-around-statements.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 5]
```diff
@@ -0,0 +1,823 @@
...
+    "registry",
+    //NOTE: use hmassD0 for normalisation, and hmass2DCorrelationPairs for 2D-sideband-subtraction purposes
+    {{"hmass", "D0,D0bar candidates;inv. mass (#pi K) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{200, 1., 3.}}}},
+     {"hmassD0", "D0,D0bar candidates;inv. mass D0 only (#pi K) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{200, 1., 3.}}}},
+     {"hmassD0bar", "D0,D0bar candidates;inv. mass D0bar only (#pi K) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{200, 1., 3.}}}},
+     {"hmass2DCorrelationPairs", "D0,D0bar candidates 2D;inv. mass D0 only (#pi K) (GeV/#it{c}^{2});inv. mass D0bar only (#pi K) (GeV/#it{c}^{2});entries", {HistType::kTH2F, {{200, 1., 3.}, {200, 1., 3.}}}},
+     {"hptcand", "D0,D0bar candidates;candidate #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hptprong0", "D0,D0bar candidates;prong 0 #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hptprong1", "D0,D0bar candidates;prong 1 #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}}},
+     {"hselectionstatus", "D0,D0bar candidates;selection status;entries", {HistType::kTH1F, {{5, -0.5, 4.5}}}},
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Better to stick to the camelCase convention.
(We have to fix it in the existing tasks so it's better if the new code is already ok.)

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 6]
```diff
@@ -0,0 +1,823 @@
...
+     {"hDeltaEtaVsPt", "D0,D0bar candidates;#it{p}_{T}^{D0};#it{p}_{T}^{D0bar};#it{#eta}^{D0bar}-#it{#eta}^{D0};entries", {HistType::kTH3F, {{100, 0., 10.}, {100, 0., 10.}, {200, -10., 10.}}}},
+     {"hDeltaPhiVsPt", "D0,D0bar candidates;#it{p}_{T}^{D0};#it{p}_{T}^{D0bar};#it{#varphi}^{D0bar}-#it{#varphi}^{D0};entries", {HistType::kTH3F, {{100, 0., 10.}, {100, 0., 10.}, {32, -o2::constants::math::PI / 2., 3. * o2::constants::math::PI / 2.}}}},
+     {"hDeltaPtDDbar", "D0,D0bar candidates;#it{p}_{T}^{D0bar}-#it{p}_{T}^{D0};entries", {HistType::kTH1F, {{144, -36., 36.}}}},
+     {"hDeltaPtMaxMin", "D0,D0bar candidates;#it{p}_{T}^{max}-#it{p}_{T}^{min};entries", {HistType::kTH1F, {{72, 0., 36.}}}},
+     {"hDDbarVsEtaCut", "D0,D0bar pairs vs #eta cut;#eta_{max};entries", {HistType::kTH2F, {{(int)(10 * maxEtaCut), 0., maxEtaCut}, {(int)(2 * PtThresholdForMaxEtaCut), 0., PtThresholdForMaxEtaCut}}}}}}; //don't modify the binning from here: act on maxEtaCut and PtThresholdForMaxEtaCut instead
+
+  Configurable<int> d_selectionFlagD0{"d_selectionFlagD0", 1, "Selection Flag for D0"};
+  Configurable<int> d_selectionFlagD0bar{"d_selectionFlagD0bar", 1, "Selection Flag for D0bar"};
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Make it camelCase here as well.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 7]
```diff
@@ -0,0 +1,823 @@
...
+
+void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
+{
+  ConfigParamSpec optionDoMC{"doMC", VariantType::Bool, false, {"Run MC-dedicated tasks."}};
+  workflowOptions.push_back(optionDoMC);
+}
+
+#include "Framework/runDataProcessing.h"
+
+double getDeltaPhi(double phiD, double phiDbar)
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx:48]: Add doxygen documentation.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 8]
```diff
@@ -0,0 +1,823 @@
...
+
+  if (dPhi < -o2::constants::math::PI / 2.)
+    dPhi = dPhi + 2. * o2::constants::math::PI;
+  if (dPhi > 3. * o2::constants::math::PI / 2.)
+    dPhi = dPhi - 2. * o2::constants::math::PI;
+
+  return dPhi;
+}
+
+double evaluatePhiByVertex(double xVertex1, double xVertex2, double yVertex1, double yVertex2)
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx:64]: Add doxygen documentation.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 9]
```diff
@@ -0,0 +1,823 @@
...
+            continue;
+          }
+          if (candidate2.isSelD0bar() >= d_selectionFlagD0bar) {
+            if (cutEtaCandMax >= 0. && std::abs(candidate2.eta()) > cutEtaCandMax)
+              continue;
+            if (cutPtCandMin >= 0. && std::abs(candidate2.pt()) < cutPtCandMin)
+              continue;
+            //Excluding trigger self-correlations (possible in case of both mass hypotheses accepted)
+            if (candidate1.mRowIndex == candidate2.mRowIndex)
+              continue;
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Add readability-braces-around-statements.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 10]
```diff
@@ -0,0 +1,823 @@
...
+            double eta1 = candidate1.eta(), eta2 = candidate2.eta(), pt1 = candidate1.pt(), pt2 = candidate2.pt(), phi1 = candidate1.phi(), phi2 = candidate2.phi();
+            registry.fill(HIST("hmass2DCorrelationPairs"), InvMassD0(candidate1), InvMassD0bar(candidate2));
+            registry.fill(HIST("hDeltaEtaPtInt"), eta2 - eta1);
+            registry.fill(HIST("hDeltaPhiPtInt"), getDeltaPhi(phi2, phi1));
+            registry.fill(HIST("hCorrel2DPtInt"), getDeltaPhi(phi2, phi1), eta2 - eta1);
+            registry.fill(HIST("hDeltaEtaVsPt"), pt1, pt2, eta2 - eta1);
+            registry.fill(HIST("hDeltaPhiVsPt"), pt1, pt2, getDeltaPhi(phi2, phi1));
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: `eta2 - eta1` and `getDeltaPhi(phi2, phi1)` are used three times. It would make sense to use variables for them.
(I don't know if the compiler is smart enough to figure it out, but it would be better for code readability as well.)

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 11]
```diff
@@ -0,0 +1,823 @@
...
+            registry.fill(HIST("hDeltaPtDDbar"), pt2 - pt1);
+            registry.fill(HIST("hDeltaPtMaxMin"), std::max(pt2, pt1) - std::min(pt2, pt1));
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Isn't this equivalent to `std::abs(pt2 - pt1)`?

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 12]
```diff
@@ -0,0 +1,823 @@
...
+     {"hDDbarVsEtaCut", "D0,D0bar pairs vs #eta cut;#eta_{max};entries", {HistType::kTH2F, {{(int)(10 * maxEtaCut), 0., maxEtaCut}, {(int)(2 * PtThresholdForMaxEtaCut), 0., PtThresholdForMaxEtaCut}}}}}}; //don't modify the binning from here: act on maxEtaCut and PtThresholdForMaxEtaCut instead
+
+  Configurable<int> d_selectionFlagD0{"d_selectionFlagD0", 1, "Selection Flag for D0"};
+  Configurable<int> d_selectionFlagD0bar{"d_selectionFlagD0bar", 1, "Selection Flag for D0bar"};
+  Configurable<double> cutEtaCandMax{"cutEtaCandMax", -1., "max. cand. pseudorapidity"};
+  Configurable<double> cutPtCandMin{"cutPtCandMin", -1., "min. cand. pT"};
+
+  Filter filterSelectCandidates = (aod::hf_selcandidate_d0::isSelD0 >= d_selectionFlagD0 || aod::hf_selcandidate_d0::isSelD0bar >= d_selectionFlagD0bar);
+
+  void process(aod::Collision const& collision, soa::Filtered<aod::Big2Prong> const& candidates, aod::BigTracks const& tracks)
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: The track table does not seem to be used anywhere.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 13]
```diff
@@ -0,0 +1,823 @@
...
+  Configurable<double> cutPtCandMin{"cutPtCandMin", -1., "min. cand. pT"};
+
+  void process(aod::McCollision const& mccollision, soa::Join<aod::McParticles, aod::HfCandProng2MCGen> const& particlesMC)
+  {
+    registry.fill(HIST("hMCEvtCount"), 0);
+    int counterccbar = 0, counterccbarPreEtasel = 0;
+
+    //loop over particles at MC gen level
+    for (auto& particle1 : particlesMC) {
+      if (std::abs(particle1.pdgCode()) == 4) { //c or cbar quark found
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Why don't you just continue if it isn't 4?

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 14]
```diff
@@ -0,0 +1,823 @@
...
+        int partMothPDG = particlesMC.iteratorAt(particle1.mother0()).pdgCode();
+        if (std::abs(partMothPDG) != 4)
+          counterccbarPreEtasel++; //count c or cbar when it doesn't come from itself via fragmenatation (c->c+X)
+        if (cutEtaCandMax >= 0. && std::abs(particle1.eta()) > cutEtaCandMax)
+          continue;
+        if (cutPtCandMin >= 0. && std::abs(particle1.pt()) < cutPtCandMin)
+          continue;
+        registry.fill(HIST("hptcandMCGen"), particle1.pt());
+        registry.fill(HIST("hEtaMCGen"), particle1.eta());
+        registry.fill(HIST("hPhiMCGen"), particle1.phi());
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Do I understand correctly that (only) these histograms should be filled for each c(bar) *and* each fragmentation stage?

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 15]
```diff
@@ -0,0 +1,823 @@
...
+        double pt1 = candidate1.pt(), phi1Std = candidate1.phi();
+        double phi1ByVtx = evaluatePhiByVertex(xPrimaryVertex, candidate1.xSecondaryVertex(), yPrimaryVertex, candidate1.ySecondaryVertex());
+        registry.fill(HIST("hPhiStdPhi"), phi1Std, pt1);
+        registry.fill(HIST("hPhiByVtxPhi"), phi1ByVtx, pt1); //trick to have correct Phi range
+        registry.fill(HIST("hPhiDifferenceTwoMethods"), getDeltaPhi(phi1ByVtx, phi1Std), pt1);
+
+        //get corresponding gen-level D0, if exists, and evaluate gen-rec phi-difference with two approaches
+        if (std::abs(candidate1.flagMCMatchRec()) == 1 << D0ToPiK) {                                                     //ok to keep both D0 and D0bar
+          int indexGen = RecoDecay::getMother(particlesMC, candidate1.index0_as<aod::BigTracksMC>().label(), 421, true); //MC-gen corresponding index for MC-reco candidate
+          if (indexGen > 0) {
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Why `if (indexGen > 0)`?

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 16]
```diff
@@ -0,0 +1,823 @@
...
+            double phi1Gen = particlesMC.iteratorAt(indexGen).phi();
+            registry.fill(HIST("hDifferenceGenPhiStdPhi"), getDeltaPhi(phi1Std, phi1Gen), pt1);
+            registry.fill(HIST("hDifferenceGenPhiByVtxPhi"), getDeltaPhi(phi1ByVtx, phi1Gen), pt1);
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Shouldn't resolution be constrained to range [-π, +π]?

**Review by vkucera — CHANGES_REQUESTED** [review]
Most comments on the first task apply to the others too.
The tasks share many hard-coded constants which should be defined only once.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 2 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Done

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 1 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Done, though I'm waiting for the merging of the #5521 pull request

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 4 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Done (also for the other equivalent cases)

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 5 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Ah ok, I had noticed that these names was not following the convention, but since was present in other tasks I hadn't modified them in mine. Done now

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 6 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Done, same comment as above

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 9 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Done

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 8 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx:64]: Done

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 7 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx:48]: Done

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 10 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Ok, done

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 11 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: You are right, substituted

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 12 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: You are right: this was needed in the previous version, before the grouping of candidates with collisions, to ensure that D0 mesons were correlated with D0bar mesons from the same event (and the collision was tracked via the information stored in the daughter track table), but after implementing the grouping that check has been removed, and the track table is not used anymore. I'm removing it.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 13 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: I modified as you proposed since it's cleaner

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 14 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Ah thanks, this is a mistake indeed: it's due to a leftover from a previous version, where I had two version of these plots (with and without inclusion of fragmentation quarks) to check the effects on those distributions from double-countings due to fragmentation.
I deleted one version after doing the check, removed the suffix from the histogram names, but left the 'wrong' version (before fragmentation check).
Also, at this point it's useless to have the instruction that skips the fragmentation (L691-692 of old commit) so late in the task, with a couple of if instructions before, so I moved that instruction at the beginning, and removed the if's.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 15 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: What I want to do is to enaure that what follows is done only if a gen-reco match is ensured, but looking how the returned variable is initialised in RecoDecay::getMother, I think I should set the condition to > -1, rather than 0 (this is what I'm doing now, please correct me if not ok)

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 16 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: I preferred having a full version with positives and negatives which acted also as control plot in case of spotting potential asymmetries (though remote), then one can always reflect while processing the plot.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 3 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: 1) Yes because cutEtaCandMax is a selection done on the candidates (which at some points could be changed to a y selection), while maxEtaCut defines the limit of the scan for the potential detector acceptance
2) so I was initially in doubt, but I think that with values large enough, as the current, they can be kept as constant (modified it, but moved outside the tasks to avoid multiple definition)
3) yes right, it's a leftover from a previous version, with dynamic increment depending on those variables, I forgot changing the comment
4) ok, modified

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 17]
```diff
@@ -670,50 +703,58 @@ struct TaskCCbarCorrelationMCGen {
...
-          counterccbar++; //count if c or cbar don't come from themselves during fragmenatation (after kinematic selection)
-
-        //c-cbar correlation dedicated section
-        //if it's c, search for cbar and evaluate correlations.
-        if (particle1.pdgCode() == 4) {
-          //check whether mothers of quark c are still '4' particles - in that case the c quark comes from its own fragmentation, skip it
-          if (partMothPDG == 4)
-            continue;
-          registry.fill(HIST("hcountD0triggersMCGen"), 0); //to count trigger c quark (for normalisation)
+      if (!std::abs(particle1.pdgCode()) == 4) { //search c or cbar particles
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: This is not gonna work. Should be:
```c++
if (std::abs(particle1.pdgCode()) != 4) {
```

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 15 above)
Comment (vkucera) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: The gen-rec match is ensured already by requiring:
```c++
if (std::abs(candidate1.flagMCMatchRec()) == 1 << D0ToPiK) {   
```
Therefore `getMother` will always return a valid index.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 16 above)
Comment (vkucera) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: The point here is that `getDeltaPhi` returns angles in range [-π/2, 3*π/2], which makes no sense for resolution.
Resolution expresses how far the rec. value is from the gen. value and that can never be larger than π for an angle.

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 18]
```diff
@@ -670,50 +703,58 @@ struct TaskCCbarCorrelationMCGen {
...
+            double eta1 = particle1.eta(), eta2 = particle2.eta(), pt1 = particle1.pt(), pt2 = particle2.pt(), phi1 = particle1.phi(), phi2 = particle2.phi();
+            double deltaEta = eta2 - eta1;
+            double deltaPhi = getDeltaPhi(phi2, phi1);
+            registry.fill(HIST("hDeltaEtaPtIntMCGen"), deltaEta);
+            registry.fill(HIST("hDeltaPhiPtIntMCGen"), deltaPhi);
+            registry.fill(HIST("hCorrel2DPtIntMCGen"), deltaPhi, deltaEta);
+            registry.fill(HIST("hDeltaEtaVsPtMCGen"), pt1, pt2, deltaEta);
+            registry.fill(HIST("hDeltaPhiVsPtMCGen"), pt1, pt2, deltaPhi);
+            registry.fill(HIST("hDeltaPtDDbarMCGen"), pt2 - pt1);
+            registry.fill(HIST("hDeltaPtMaxMinMCGen"), std::max(particle2.pt(), particle1.pt()) - std::min(particle2.pt(), particle1.pt()));
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: `std::abs`

**Review by vkucera — CHANGES_REQUESTED** [review]

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 15 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Ok, removed the condition

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 16 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Ah I misread the comment (read [0,pi] instead). Ok then, I modified the code accordingly

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 17 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Fixed

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 18 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Changed as done for previous occurrences

Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Committed updated version which satisfies the new requests for adaptAnalysisTask (#PR5558).
No modifications are needed for the charge()->sign() update (#PR5611)

**Comment by fcolamar** [comment]
Hello all, is there any further comment on this PR?
Thanks!

**Comment by vkucera** [comment]
Hi @fcolamar , your branch is 247 commits behind `dev`. Could you please update it and check that your code works fine with the latest O2?

**Comment by fcolamar** [comment]
Hi Vit, I updated the branch with the new dev commits, the only needed modification was an update to the CMakeLists.txt which was indeed modified by Fabio Catalano this morning. Tried to compile and it works fine (I couldn't test the code yet due to errors when running the validation framework, though I don't expect problems)

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 19]
```diff
@@ -0,0 +1,875 @@
...
+  Filter filterSelectCandidates = (aod::hf_selcandidate_d0::isSelD0 >= dSelectionFlagD0 || aod::hf_selcandidate_d0::isSelD0bar >= dSelectionFlagD0bar);
+
+  void process(aod::Collision const& collision, soa::Filtered<aod::Big2Prong> const& candidates)
+  {
+    for (auto& candidate1 : candidates) {
+      if (cutEtaCandMax >= 0. && std::abs(candidate1.eta()) > cutEtaCandMax) {
+        continue;
+      }
+      if (cutPtCandMin >= 0. && std::abs(candidate1.pt()) < cutPtCandMin) {
+        continue;
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Absolute value of pT?

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 20]
```diff
@@ -0,0 +1,875 @@
...
+/// Returns phi of candidate/particle evaluated from x and y components of segment connecting primary and secondary vertices
+///
+double evaluatePhiByVertex(double xVertex1, double xVertex2, double yVertex1, double yVertex2)
+{
+  return RecoDecay::Phi(xVertex2 - xVertex1, yVertex2 - yVertex1);
+}
+
+/// definition of variables for D0D0bar pairs vs eta acceptance studies (hDDbarVsEtaCut, in data-like, MC-reco and MC-kine tasks)
+const double maxEtaCut = 5.;
+const double PtThresholdForMaxEtaCut = 10.;
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: `ptThresholdForMaxEtaCut`

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 21]
```diff
@@ -0,0 +1,875 @@
...
+const double incrementEtaCut = 0.1;
+const double incrementPtThreshold = 0.5;
+const double epsilon = 10E-5;
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Is it supposed to be 1e-4?

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 22]
```diff
@@ -0,0 +1,875 @@
...
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file taskD0D0barCorrelation.cxx
+/// \brief D0-D0bar analysis task - data-like, MC-reco and MC-kine analyses. For ULS and LS pairs
+///
+/// \author Fabio Colamaria <fabio.colamaria@ba.infn.it>, INFN Bari
+
+#include <cmath>
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Why is this needed?

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx [hunk 23]
```diff
@@ -0,0 +1,875 @@
...
+
+  Produces<aod::HF2ProngCollis> create2ProngIndexCollColumn;
+  void process(aod::HfCandProng2 const& candidates, aod::Tracks const& tracks)
+  {
+    for (auto& candidate : candidates) {
+      int indexColl = candidate.index0_as<aod::Tracks>().collisionId(); //takes index of collision from first D daughter
+      create2ProngIndexCollColumn(indexColl);
+    }
+  }
+};
```
Comment (vkucera) [inline Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: This should be removed once https://github.com/AliceO2Group/AliceO2/pull/5469 gets merged.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 20 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Modified

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 21 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Modified to 1E-5

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 22 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Removed (leftover from previous version with math operations that were directly implemented here)

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 23 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Sure, for now I left it in place

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 19 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Right, copy-paste from above selection on eta done without thinking... Removed the abs (here and later on)

**Comment by fcolamar** [comment]
Draft PR ready for new review

File: Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx (hunk 23 above)
Comment (fcolamar) [reply Analysis/Tasks/PWGHF/taskD0D0barCorrelation.cxx]: Since #5469 was merged this morning, I adapted the task and committed the updated version

**Comment by fcolamar** [comment]
Adapted to merging of #5469 PR, ready for review

**Comment by fcolamar** [comment]
@ginnocen @vkucera I've just committed the new version of the D0D0bar code, with the full restructuring as discussed during yesterday's meeting

**Comment by vkucera** [comment]
@fcolamar Shall we then close this PR in favour of https://github.com/AliceO2Group/AliceO2/pull/6124?

**Comment by fcolamar** [comment]
Hi @vkucera, in my opinion is would be better, since the new PR #6124 implements what added in this PR and on top of that adds the D+D- correlations.
I had left this one open just in case you preferred going step by step
